File: src/__tests__/_canonical-bytes.ts
import { expect } from "vitest";

////////////////////////////////
//  Canonical signing-bytes assertion
//
//  A cross-runtime signing vector pins its canonical preimage in three encodings
//  (utf8 / hex / base64) so the Kotlin twin is held byte-exact against node:crypto
//  from whichever encoding it reproduces. Asserting all three per op is the same
//  three lines repeated; this helper folds them into one call. Not a `.test.ts`
//  file, so vitest does not collect it as a suite (tsc still type-checks it).

/** The triple-encoded canonical preimage a cross-runtime signing vector records. */
export interface CanonicalBytesVec {
	signingBytes: string;
	signingBytesHex: string;
	signingBytesBase64: string;
}

/** Assert `actual` reproduces the vector's canonical signing bytes in all three
 * recorded encodings (utf8, hex, base64). */
export function assertCanonicalBytes(actual: Buffer, vec: CanonicalBytesVec): void {
	expect(actual.toString("utf8")).toBe(vec.signingBytes);
	expect(actual.toString("hex")).toBe(vec.signingBytesHex);
	expect(actual.toString("base64")).toBe(vec.signingBytesBase64);
}

File: android/app/src/test/java/com/atelier_nyaarium/switchboard/crypto/CanonicalBytes.kt
package com.atelier_nyaarium.switchboard.crypto

import kotlinx.serialization.json.JsonObject
import kotlinx.serialization.json.jsonPrimitive
import org.junit.Assert.assertEquals
import java.util.Base64

/**
 * Asserts a signing-bytes preimage reproduces a cross-runtime vector's canonical
 * encodings (utf8 / hex / base64). The vector pins all three so the Kotlin twin is
 * held byte-exact against node:crypto; folding the per-encoding assertions into one
 * call keeps the per-op suites free of the repeated triple. Mirrors the TS
 * assertCanonicalBytes helper.
 */
object CanonicalBytes {
	/** `vec` is a vector entry (its `value`/`signature` siblings are ignored here);
	 * this checks `bytes` against its `signingBytes` / `signingBytesHex` /
	 * `signingBytesBase64` fields. */
	fun assertCanonicalBytes(bytes: ByteArray, vec: JsonObject) {
		assertEquals(vec["signingBytes"]!!.jsonPrimitive.content, bytes.toString(Charsets.UTF_8))
		assertEquals(vec["signingBytesHex"]!!.jsonPrimitive.content, bytes.joinToString("") { "%02x".format(it) })
		assertEquals(vec["signingBytesBase64"]!!.jsonPrimitive.content, Base64.getEncoder().encodeToString(bytes))
	}
}
